Simply SVGInlineTextBox::mapStartEndPositionsIntoFragmentCoordinates

Ahmad-S792 opened the pull request:
#### 8e69bb72e0eeddcb4aab6d5c107b986186b61a98
<pre>
Simply SVGInlineTextBox::mapStartEndPositionsIntoFragmentCoordinates

<a href="https://bugs.webkit.org/show_bug.cgi?id=262488">https://bugs.webkit.org/show_bug.cgi?id=262488</a>

Reviewed by Said Abou-Hallawa.

Reduce SVGInlineTextBox::mapStartEndPositionsIntoFragmentCoordinates to
a simple offset/base adjustment and an intersection (which is what it
did - just in an unnecessarily verbose way.)

* Source/WebCore/rendering/svg/SVGInlineTextBox.cpp:
(SVGInlineTextBox::mapStartEndPositionsIntoFragmentCoordinates):

Canonical link: <a href="https://commits.webkit.org/269425@main">https://commits.webkit.org/269425@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b4d27183c4fa7b05b324e25f8078dba75d20cd46

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22523 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/180 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23606 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24430 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20852 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22783 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/1249 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23056 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21837 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22763 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19554 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25282 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20413 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/26680 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20658 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24500 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/82 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5367 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/129 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/124 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/rendering/svg/SVGInlineTextBox.cpp:
```diff
@@ -411,32 +412,17 @@ TextRun SVGInlineTextBox::constructTextRun(const RenderStyle& style, const SVGTe
-    }
+    // Compute positions relative to the fragment.
+    startPosition -= fragmentOffsetInBox;
+    endPosition -= fragmentOffsetInBox;
```
This calculation will overflow if `startPosition` or `endPosition` is less than `fragmentOffsetInBox` because all of them are `unsigned`. The original code was more correct than this new one because it has early returns to check these cases. However you can use something like this if you want:
```
bool SVGInlineTextBox::mapStartEndPositionsIntoFragmentCoordinates(const SVGTextFragment& fragment, unsigned& startPosition, unsigned& endPosition) const
{
    unsigned startFragment = fragment.characterOffset - start();
    unsigned endFragment = startFragment + fragment.length;

    // Find intersection between the intervals: [startFragment..endFragment) and [startPosition..endPosition)
    startPosition = std::max(startFragment, startPosition);
    endPosition = std::min(endFragment, endPosition);

    if (startPosition >= endPosition)
        return false;

    startPosition -= startFragment;
    endPosition -= startFragment;
    return true;
}
```

Ahmad-S792 replied:
Pushed based on above. Thanks! Now will let EWS run it. :-)

Ahmad-S792 replied:
@shallawa - Thanks for your help. Seems to work fine and got bit simplification in.

shallawa commented on Source/WebCore/rendering/svg/SVGInlineTextBox.cpp:
```diff
@@ -411,31 +411,19 @@ TextRun SVGInlineTextBox::constructTextRun(const RenderStyle& style, const SVGTe
-        endPosition -= offset;
-    }
-
-    ASSERT_WITH_SECURITY_IMPLICATION(startPosition < endPosition);
```
I think it is important to mention that there are there coordinates system for the SVG text:

1. Line coordinates
2. InlineTextBox coordinates
3. Fragment coordinates

`SVGInlineTextBox` and `SVGTextFragment` are both in the _Line coordinates_. The `startPosition` and `endPosition` are both in _InlineTextBox coordinates_. And this function maps them to the _Fragment coordinates_.

shallawa commented on Source/WebCore/rendering/svg/SVGTextQuery.cpp:
```diff
@@ -137,63 +137,53 @@ bool SVGTextQuery::executeQuery(Data* queryData, ProcessTextFragmentCallback fra
+    // Reuse the same logic used for text selection & painting, to map our
+    // query start/length into start/endPositions of the current text fragment.
     if (!queryData->textBox->mapStartEndPositionsIntoFragmentCoordinates(fragment, startPosition, endPosition))
         return false;
```
This will map the `startPosition` and `endPosition` to the _Fragment coordinates_ before expanding them for the ligatures.

shallawa commented on Source/WebCore/rendering/svg/SVGTextQuery.cpp:
```diff
@@ -137,63 +137,53 @@ bool SVGTextQuery::executeQuery(Data* queryData, ProcessTextFragmentCallback fra
-    startPosition = std::max<unsigned>(startPosition, fragmentStartInBox);
-    endPosition = std::min<unsigned>(endPosition, fragmentEndInBox);
+    startPosition -= boxStart;
+    endPosition -= boxStart;
```
This will map the `startPosition` and `endPosition` from the _Line coordinates_ to the _InlineTextBox coordinates_.

shallawa commented on Source/WebCore/rendering/svg/SVGTextQuery.cpp:
```diff
@@ -137,63 +137,53 @@ bool SVGTextQuery::executeQuery(Data* queryData, ProcessTextFragmentCallback fra
     if (!queryData->textBox->mapStartEndPositionsIntoFragmentCoordinates(fragment, startPosition, endPosition))
         return false;
 
+    modifyStartEndPositionsRespectingLigatures(queryData, fragment, startPosition, endPosition);
```
I do not think this is correct. I think this step has to be followed by intersecting the `startPosition` and  `endPosition` with the fragment boundaries.

shallawa commented on Source/WebCore/rendering/svg/SVGTextQuery.cpp:
```diff
@@ -137,63 +137,53 @@ bool SVGTextQuery::executeQuery(Data* queryData, ProcessTextFragmentCallback fra
 
     unsigned textMetricsOffset = fragment.metricsListOffset;
+    unsigned fragmentOffset = 0;
+    unsigned fragmentEnd = fragment.length;
```
Now `startPosition` and  `endPosition` are both in _Fragment coordinates_. But I do not understand why this is more simplified than the original code.

I think it is clearer to have `SVGInlineTextBox` and `SVGTextFragment` in _Line coordinates_ and have the calculations of `startPosition` and `endPosition` in _InlineTextBox coordinates_.  The conversion to the _Fragment coordinates_ and intersecting them with the Fragment boundaries should the last step. But to run some calculations in _InlineTextBox coordinates_ and others in _Fragment coordinates_ is confusing.

Ahmad-S792 committed:
Simply SVGInlineTextBox::mapStartEndPositionsIntoFragmentCoordinates

https://bugs.webkit.org/show_bug.cgi?id=262488

Reviewed by Said Abou-Hallawa.

Reduce SVGInlineTextBox::mapStartEndPositionsIntoFragmentCoordinates to
a simple offset/base adjustment and an intersection (which is what it
did - just in an unnecessarily verbose way.)

* Source/WebCore/rendering/svg/SVGInlineTextBox.cpp:
(SVGInlineTextBox::mapStartEndPositionsIntoFragmentCoordinates):

Canonical link: https://commits.webkit.org/269425@main
